python/pyarrow/pandas_compat.py

```diff
@@ -950,7 +950,7 @@ def _reconstruct_index(table, index_descriptors, all_columns, types_mapper=None)
         index = index_arrays[0]
         if not isinstance(index, pd.Index):
             # Box anything that wasn't boxed above
-            index = pd.Index(index, name=index_names[0])
+            index = pd.Index(index.infer_objects(), name=index_names[0])
```

Comment:
Do you know for which test this is needed? (I mean, which test triggers the warning here) 
I am wondering what `index` would be then, because I would think we always already created a proper pandas object.

Reply:
It is `test_roundtrip_empty_table_with_extension_dtype_index`.

Reply:
OK, so putting a breakpoint at the above location and then running that test, I see:

```
(Pdb) index
0    (0, 1]
1    (1, 2]
2    (2, 3]
dtype: object
(Pdb) index.values
array([Interval(0, 1, closed='right'), Interval(1, 2, closed='right'),
       Interval(2, 3, closed='right')], dtype=object)
(Pdb) pd.Index(index, name=index_names[0])
IntervalIndex([(0, 1], (1, 2], (2, 3]], dtype='interval[int64, right]')
```

So `index` is an object-dtype Series with scalar Interval objects, and then converting that to an Index on the line above currently infers that it are interval objects, and converts that to interval dtype, while in the future it will preserve the object dtype.

Now, I do wonder why we have an object dtype series of interval objects, though. Because we created that data (converted from a column in the pyarrow table), so I would expect that we create proper interval dtype to start with.

Reply:
The index object is created in `_extract_index_level`, and it seems that for interval dtype, the `.values` call is causing conversion to object dtype:

```
(Pdb) l
971  	        return result_table, None, None
972  	
973  	    pd = _pandas_api.pd
974  	
975  	    col = table.column(i)
976  ->	    values = col.to_pandas(types_mapper=types_mapper).values
977  	
978  	    if hasattr(values, 'flags') and not values.flags.writeable:
979  	        # ARROW-1054: in pandas 0.19.2, factorize will reject
980  	        # non-writeable arrays when calling MultiIndex.from_arrays
981  	        values = values.copy()

(Pdb) col.to_pandas()
0    (0, 1]
1    (1, 2]
2    (2, 3]
Name: __index_level_0__, dtype: interval
(Pdb) col.to_pandas().values
array([Interval(0, 1, closed='right'), Interval(1, 2, closed='right'),
       Interval(2, 3, closed='right')], dtype=object)
```

The reason we do the `.values` is because we check whether the data is writeable for old pandas versions. This if block can nowadays maybe be removed, and then we don't need to conversion of a Series to an array.

Reply:
I think we create a series here: https://github.com/apache/arrow/blob/ac50918a63f13088a5ea9e6c27a9fadbbb19d53f/python/pyarrow/pandas_compat.py#L983

Reply:
Sorry, I have missed your last comment (the page didn't refresh because I was writing at the same time, I guess). Will remove the if block and check if it works 👍

Reply:
https://github.com/apache/arrow/pull/39609/commits/01980c1baa76786f6bbae73d16ec04918c4b502b

Reply:
Nice cleanup!